src/ucs/sys/topo.c

```diff
@@ -19,9 +53,125 @@ ucs_sys_bus_id_t ucs_sys_bus_id_unknown = { .domain   = 0xffff,
                                             .function = 0xff
                                           };
 
+static ucs_bus_id_bit_rep_t ucs_topo_get_bus_id_bit_repr(const ucs_sys_bus_id_t *bus_id)
+{
+    ucs_bus_id_bit_rep_t bit_rep = 0;
+
+    bit_rep |= bus_id->domain;
+    bit_rep <<= 24;
+    bit_rep |= bus_id->bus;
+    bit_rep <<= 16;
+    bit_rep |= bus_id->slot;
+    bit_rep <<= 8;
+    bit_rep |= bus_id->function;
+
+    return bit_rep;
+}
+
+void ucs_topo_init()
+{
+    if (ucs_topo_initialized) return;
+
+    ucs_recursive_spinlock_init(&ucs_topo_sys_dev_to_bus_lock, 0);
+    ucs_recursive_spinlock_init(&ucs_topo_bus_to_sys_dev_lock, 0);
+    kh_init_inplace(bus_to_sys_dev, &ucs_topo_bus_to_sys_dev_hash);
+    ucs_topo_sys_dev_to_bus_lookup.count = 0;
+
+    ucs_topo_initialized = 1;
+}
+
+void ucs_topo_cleanup()
+{
+    ucs_status_t status;
+    ucs_sys_device_t sys_dev;
+
+    if (!ucs_topo_initialized) return;
+
+    kh_foreach_value(&ucs_topo_bus_to_sys_dev_hash, sys_dev, {
+        ucs_debug("sys_device_t %u not released", sys_dev);
```

Comment:
who is responsible to release it?

Reply:
No one is releasing this at the moment so I left it to `kh_destroy`. Entries are added into the hashmap each time a new device is discovered but I can't think of a good reason to remove it until `topo_cleanup` is called.

Reply:
so why need this debug message?

Reply:
Kept it for tracking purposes.